client/src/Security/LoginFormAuthenticator.php

```diff
@@ -72,7 +71,13 @@ public function authenticate(Request $request): Passport
 
     public function onAuthenticationSuccess(Request $request, TokenInterface $token, string $firewallName): ?Response
     {
-        $redirectUrl = $this->redirector->getFirstPageAfterLogin($request->getSession());
+        $redirectUrl = (
+            $request->query->get('lastPage') &&
+            RouteValidator::validateRoute($this->router, $request->query->get('lastPage'))
+        ) ?
+            urldecode($request->query->get('lastPage')) :
```

Comment:
url-decoding here, but not in what you pass to the validator? So the validator is working on an encoded url? Might potentially lead to unexpected behaviour.